providers/client_socket.js

```diff
@@ -122,6 +127,15 @@ ClientSocket.prototype.write = function(data) {
   this.outputStream.write(stringData, stringData.length);
 };
 
+ClientSocket.prototype.pause = function() {
+  this.paused = true;
+};
+
+ClientSocket.prototype.resume = function() {
+  this.paused = false;
+  this.rawInputStream.asyncWait(this.inputStreamCallback, 0, 0, mainThread);
```

Comment:
seems plausible. are there flags we want to the input stream https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsITransport#openInputStream() that would make the back pressure more effective (by e.g. restricting how much can be in queue, which firefox maybe uses to define its tcp window size?)